crates/server/src/config/mod.rs

```diff
@@ -318,3 +334,34 @@ impl ZoneConfig {
         &self.keys
     }
 }
+
+#[cfg(feature = "dnssec")]
+/// NSEC3 related configuration.
+#[derive(Deserialize, PartialEq, Eq, Debug, Clone, Default)]
+pub struct Nsec3Config {
+    /// Algorithm used to create the hashed NSEC3 owner names
+    #[serde(default)]
+    pub hash_algorithm: Nsec3HashAlgorithm,
+    /// Salt used to create the hashed NSEC3 owner names
+    #[serde(default)]
+    pub salt: Vec<u8>,
+    /// Number of iterations used to create the hashed NSEC3 owner names
+    #[serde(default)]
+    pub iterations: u16,
+}
+
+/// The kind of non-existence proof provided by the server.
+#[derive(Debug, Default, Deserialize, Clone, Copy, PartialEq, Eq)]
+#[serde(rename_all = "lowercase")]
+pub enum NxProof {
+    #[cfg(feature = "dnssec")]
+    #[cfg_attr(feature = "dnssec", default)]
+    /// Use NSEC
+    Nsec,
+    #[cfg(feature = "dnssec")]
+    /// Use NSEC3
+    Nsec3,
+    #[cfg_attr(not(feature = "dnssec"), default)]
+    /// Do not provide non-existence proofs
+    None,
```

Comment:
> Why not model this as `Option<NxProof>`?

Two reasons:
- Because `NxProof` is shorter
- `NxProof::None` is clearly documented as the option where no non-existence proof is provided instead of having to wonder if `Option::None` means that no non-existence proof will be provided or that the default behavior will be used (which changes if the `dnssec` feature is enabled)

Reply:
I don't find the "shorter" reason very compelling -- it's more important to me that we have a canonical way of modeling what's going on rather than fewer characters to read/write.

> * `NxProof::None` is clearly documented as the option where no non-existence proof is provided instead of having to wonder if `Option::None` means that no non-existence proof will be provided or that the default behavior will be used (which changes if the `dnssec` feature is enabled)

Is there a difference in behavior for these two cases ("not providing a non-existence proof" vs "default behavior, with dnssec disabled)?

Reply:
disclaimer: I'm commenting with limited context as I have not read through all the changes

In principle, I see no problem in flattening `Option<TwoVariantNxProof>` into a `ThreeVariantNxProof` enum so long we are consistent everywhere else in *not* using `Option<ThreeVariantNxProof>`. I'm less worried with the `Option<ThreeVariantNxProof>` type being allowed to exist; just like I'm not worried about it being possible to write a type like `Result<Something, Option<Error>>`

by "being consistent" I mean that the `Default` implementation of `ThreeVariantNxProof` should return the `None` variant. if the field is absent in a configuration file, serde should decode that into the `ThreeVariantNxProof::None` variant. if the flag is not passed via the command line, then assume `ThreeVariantNxProof::None` etc.

similarly, we should use the same `ThreeVariantNxProof` type regardless of whether the dnssec feature is enabled or not and we should conditionally remove the variants that make no sense when dnssec is disabled

other thing to consider is whether we want to group this with other DNSSEC related settings. perhaps at some point, it would make sense to have some `Dnssec` struct that contains `TwoVariantNxProof` (and other fields) and sits behind an `Option` when it appears in a configuration struct or something like that

Reply:
> I'm less worried with the `Option<ThreeVariantNxProof>` type being allowed to exist;

I do think this can be an issue; `ThreeVariantNxProof` won't have such a verbose name in practice, so it's not obvious from reading that `Option<NxProof>` could have another `None` instead, and without context someone might assume that passing a non-`Option`-wrapped `NxProof` would always result in a proof being generated, when this is not the case because `NxProof` might be `NxProof::None`.

In other words, I have a strong preference towards `enum NxProofType { Nsec, Nsec3 }` with non-providers using `Option::<NxProofType>::None` to represent this in the type system, so basically I want to hear objections to why such a representation would cause more problems compared to the current one.

Reply:
>  so basically I want to hear objections to why such a representation would cause more problems compared to the current one.

Doing `enum NxProof { Nsec, Nsec3 }` and using `Option<NxProof>` everywhere means that if the `dnssec` feature is disabled you only have two choices for `NxProof`:
- Gate it behind `dnssec` as well: Which means that any item that contains `NxProof` needs to be gated through all the codebase
- Gate the two variants behind `dnssec`: Which means that you have an empty `enum` in your codebase which is not the most ergonomic thing to handle.

Reply:
That makes sense as reasoning. Gating the existence of `NxProof` itself would make sense to me, I'm not sure how many call-sites that would touch? It seems conceptually correct.